documentation/developer_setup.md

```diff
@@ -0,0 +1,336 @@
+
+
+## Write Tests
+
+Testing is built into the Go toolchain as well with the `testing` library. The testing infrastructure located in the `sdk/internal/recording` directory takes care of generating recordings, establishing the mode a test is being run in (options are "record" or "playback") and reading environment variables. The HTTP traffic is intercepted by a custom [test-proxy][test_proxy_docs] in both the "recording" and "playback" case to either persist or read HTTP interactions from a file. There is one small step that needs to be added to you client creation to route traffic to this test proxy. All three of these modes are specified in the `AZURE_RECORD_MODE` environment variable:
+
+| Mode | Powershell Command |
+| ---- | ------------------ |
+| record | `$ENV:AZURE_RECORD_MODE="record"` |
+| playback | `$ENV:AZURE_RECORD_MODE="playback"` |
+
+To get started first install [`docker`][get_docker]. Then to start the proxy, from the root of the repository, run the command `./eng/common/testproxy/docker-start-proxy.ps1 start`. This command will take care of pulling the pinned docker image and running it in the background.
+
+It is not required to run the test-proxy from within the docker container, but this is how the proxy is run in the Azure DevOps pipelines. If you would like to run the test-proxy in a different way check out the test-proxy [documentation][test_proxy_docs] for more information.
+
+
+### Test Mode Options
+
+There are two options for test modes: "recording" and "playback" each with their own purpose.
+
+Recording mode is for testing against a live service and 'recording' the HTTP interactions in a JSON file for use later. This is helpful for developers because not every request will have to run through the service and makes your tests run much quicker. This also allows us to run our tests in public pipelines without fear of leaking secrets to our developer subscriptions.
+
+In playback mode the JSON file that the HTTP interactions are saved to is used in place of a real HTTP call. This is quicker and is used most often for quickly verifying you did not change the behavior of your library.
+
+
+### Routing Requests to the Proxy
+
+All clients should contain an options struct as the last parameter on the constructor. In this options struct you need to have a way to provide a custom HTTP transport object. In your tests, you will replace the default HTTP transport object with a custom one in the `internal/recording` library that takes care of all the routing for you. For example, here is that code snippet in the `aztable` package:
+
+```golang
+package aztable
+
+import (
+	...
+
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/recording"
+)
+
+var pathToPackage = "sdk/data/aztables"
+
+type recordingPolicy struct {
```

Comment:
Are we requiring SDK authors to write this themselves, and if so, should we provide one via the `recording` package instead?

Reply:
Same comment about having circular dependency

Reply:
Hmm yes I forgot about that.